lib/rucio/common/test_rucio_server.py

```diff
@@ -150,3 +152,408 @@ def test_upload_download(self):
+        cmd = 'rucio download --dir /tmp/ {0}:{1} --impl {2}'.format(self.scope, tmp_file1, impl)
+        print(self.marker + cmd)
+        exitcode, out, err = execute(cmd)
+        print(out)
+        print(err)
+        # The file should be there
+        cmd = 'ls /tmp/{0}/rucio_testfile_*'.format(self.scope)
+        print(self.marker + cmd)
+        exitcode, out, err = execute(cmd)
+        print(err, out)
+        self.assertEqual(exitcode, 0)
+
+        # cleaning
+        remove('/tmp/{0}/'.format(self.scope) + tmp_file1[5:])
+        added_dids = ['{0}:{1}'.format(self.scope, did) for did in (tmp_file1)]
+        self.generated_dids += added_dids
+
+    def test_upload_download_datasets_using_impl(self):
+        """CLIENT(USER): rucio upload files to dataset/download dataset"""
+        tmp_file1 = file_generator()
+        tmp_file2 = file_generator()
+        tmp_file3 = file_generator()
+        tmp_dsn = 'tests.rucio_client_test_server_impl_check' + uuid()
+        impl = 'rucio.rse.protocols.xrootd.Default'
+
+        # Adding files to a new dataset
+        cmd = 'rucio upload --rse {0} --scope {1} --impl {2} {3} {4} {5} {1}:{6}'.format(self.rse, self.scope, impl, tmp_file1, tmp_file2, tmp_file3, tmp_dsn)
+        print(self.marker + cmd)
+        exitcode, out, err = execute(cmd)
+        print(out)
+        print(err)
+        remove(tmp_file1)
+        remove(tmp_file2)
+        remove(tmp_file3)
+        self.assertEqual(exitcode, 0)
+
+        # List the files
+        cmd = 'rucio list-files {0}:{1}'.format(self.scope, tmp_dsn)
+        print(self.marker + cmd)
+        exitcode, out, err = execute(cmd)
+        print(out)
+        print(err)
+        self.assertEqual(exitcode, 0)
+
+        # List the replicas
+        cmd = 'rucio list-file-replicas {0}:{1}'.format(self.scope, tmp_dsn)
+        print(self.marker + cmd)
+        exitcode, out, err = execute(cmd)
+        print(out)
+        print(err)
+        self.assertEqual(exitcode, 0)
+
+        # Downloading dataset
+        cmd = 'rucio download --dir /tmp/ {0}:{1} --impl {2}'.format(self.scope, tmp_dsn, impl)
+        print(self.marker + cmd)
+        exitcode, out, err = execute(cmd)
+        print(out)
+        print(err)
+        # The files should be there
+        cmd = 'ls /tmp/{0}/rucio_testfile_*'.format(tmp_dsn)
+        print(self.marker + cmd)
+        exitcode, out, err = execute(cmd)
+        print(err, out)
+        self.assertEqual(exitcode, 0)
+
+        # cleaning
+        remove('/tmp/{0}/'.format(tmp_dsn) + tmp_file1[5:])
+        remove('/tmp/{0}/'.format(tmp_dsn) + tmp_file2[5:])
+        remove('/tmp/{0}/'.format(tmp_dsn) + tmp_file3[5:])
+        added_dids = ['{0}:{1}'.format(self.scope, did) for did in (tmp_file1, tmp_file2, tmp_file3, tmp_dsn)]
+        self.generated_dids += added_dids
+
+    def test_repeat_upload_download_file_using_impl(self):
+        """CLIENT(USER): rucio repeat upload/download file"""
```

Comment:
I followed the same pattern as the one in test_bin_rucio, but I was wondering if i should change it to also check if the file is already there it will not reupload it?

Reply:
can be a separate additional test?

Reply:
@mlassnig yep okay that sounds good i'll do that